Added btcmarkets streaming code

nielsdraaisma opened the pull request.

badgerwithagun commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingMarketDataService.java:
```diff
@@ -0,0 +1,52 @@
+    return service
+        .subscribeChannel(CHANNEL_ORDERBOOK, marketId)
+        .map(node -> mapper.treeToValue(node, BTCMarketsWebSocketOrderbookMessage.class))
+        .filter(orderEvent -> orderEvent.marketId != null && orderEvent.marketId.equals(marketId))
```
could simplify this to `marketId.equals(orderEvent.marketId)` and lose the null check

nielsdraaisma replied:
Done

badgerwithagun commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java:
```diff
@@ -0,0 +1,70 @@
+import org.slf4j.LoggerFactory;
+
+public class BTCMarketsStreamingService extends JsonNettyStreamingService {
+  public static final String CHANNEL_ORDERBOOK = "orderbook";
```
Do these need to be `public`?

nielsdraaisma replied:
I've moved things out of the packages and removed  `public`.

badgerwithagun commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java:
```diff
@@ -0,0 +1,70 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BTCMarketsStreamingService extends JsonNettyStreamingService {
```
Does this class itself need to be `public`?

nielsdraaisma replied:
I've moved things out of the packages and removed  `public`.

badgerwithagun commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java:
```diff
@@ -0,0 +1,70 @@
+  @Override
+  public String getUnsubscribeMessage(String channelName) throws IOException {
+    if (channelName.startsWith(CHANNEL_ORDERBOOK)) {
+      final String[] parts = channelName.split(":");
```
Looks like there should be a `subscribedOrderbooks.remove()` here.

nit: Setting up `subscribedOrderbooks` in the `get...Message` methods seems unsafe, but I can't see an obvious cleaner way. It's not the ideal websocket API, so I'm not too bothered.

nielsdraaisma replied:
Done, there's no clean way of working around that, an unsubscribe is basically subscribing again without that channel.

nielsdraaisma commented on xchange-stream-btcmarkets/src/test/resources/log4j2.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Removed that.

nielsdraaisma commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/dto/BTCMarketsWebSocketOrderbookMessage.java:
```diff
@@ -0,0 +1,28 @@
+  public BTCMarketsWebSocketOrderbookMessage(
+      @JsonProperty("marketId") String marketId,
+      @JsonProperty("timestamp") String timestamp,
+      @JsonProperty("bids") List<List<String>> bids,
```
Done

nielsdraaisma commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/BTCMarketsStreamingAdapters.java:
```diff
@@ -0,0 +1,44 @@
+import org.knowm.xchange.dto.trade.LimitOrder;
+import org.knowm.xchange.utils.DateUtils;
+
+public class BTCMarketsStreamingAdapters {
```
Done

nielsdraaisma commented on xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingDigest.java:
```diff
@@ -0,0 +1,32 @@
+import javax.crypto.spec.SecretKeySpec;
+import org.knowm.xchange.utils.Assert;
+
+public class BTCMarketsStreamingDigest {
```
I've removed that class entirely, i'll raise a new PR later with working private channels, that's not currently used at all.

nielsdraaisma commented:
Thanks @badgerwithagun for your review, aside from some visibility issues i've addressed the issues you raised. I've moved things around a bit and removed the public visibility of some classes.
